stats/cache.go

```diff
@@ -3,79 +3,262 @@ package stats
-	// TODO (b5) - use this
-	_ = filepath.Join(c.root, filename)
+func (c *localCache) PutStats(ctx context.Context, key string, sa *dataset.Stats) (err error) {
+	var statProps, targetProps didmod.Props
+	if qfs.PathKind(key) == "local" {
+		targetProps, _ = didmod.NewProps(key)
+	}
+
+	key = c.cacheKey(key)
+	filename := c.componentFilepath(key)
+	data, err := json.Marshal(sa)
+	if err != nil {
+		return err
+	}
+
+	if int64(len(data)) > c.maxSize {
+		return fmt.Errorf("stats component size exceeds maximum size of cache")
+	}
 
-	return fmt.Errorf("not finished")
+	if err := ioutil.WriteFile(filename, data, 0644); err != nil {
+		return err
+	}
+	statProps, err = didmod.NewProps(filename)
+	if err != nil {
+		return err
+	}
+
+	c.addAndPurgeExpired(key, statProps, targetProps)
+
+	return c.writeCacheInfo()
 }
 
-// JSON gets cached byte data for a path
-func (c osCache) JSON(ctx context.Context, path string) (r io.Reader, err error) {
-	return nil, ErrCacheMiss
+// Stats gets cached byte data for a path
+func (c *localCache) GetStats(ctx context.Context, key string) (sa *dataset.Stats, err error) {
+	cacheKey := c.cacheKey(key)
+	log.Debugw("getting stats", "key", key, "cacheKey", cacheKey)
+
+	targetFileProps, exists := c.info.TargetFileProps[cacheKey]
+	if !exists {
+		return nil, ErrCacheMiss
+	}
+
+	if qfs.PathKind(key) == "local" {
+		if fileProps, err := didmod.NewProps(key); err == nil {
+			if !targetFileProps.Equal(fileProps) {
+				// note: returning ErrCacheMiss here will probably lead to re-calcualtion
+				// and subsequent overwriting by cache consumers, so we shouldn't need
+				// to proactively drop the stale cache here
+				return nil, ErrCacheMiss
+			}
+		}
+	}
+
+	f, err := os.Open(c.componentFilepath(cacheKey))
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	sa = &dataset.Stats{}
+	err = json.NewDecoder(f).Decode(sa)
+	return sa, err
 }
 
 var b32Enc = base32.NewEncoding("abcdefghijklmnopqrstuvwxyz234567").WithPadding(base32.NoPadding)
 
-// nilCache is a stand in for not having a cache
-// it only ever returns ErrNoCache
-type nilCache bool
+func (c *localCache) componentFilepath(cacheKey string) string {
+	return filepath.Join(c.root, fmt.Sprintf("%s.json", cacheKey))
+}
 
-var _ Cache = (*nilCache)(nil)
+func (c *localCache) cacheKey(key string) string {
+	return b32Enc.EncodeToString([]byte(key))
+}
 
-// PutJSON places stats in the cache, keyed by path
-func (nilCache) PutJSON(ctx context.Context, path string, r io.Reader) error {
-	return ErrNoCache
+const uintSize = 32 << (^uint(0) >> 32 & 1)
+
+func (c *localCache) addAndPurgeExpired(cacheKey string, statProps, targetProps didmod.Props) {
+	c.infoLk.Lock()
+	defer c.infoLk.Unlock()
+	log.Debugw("adding stat props", "cacheKey", cacheKey, "statProps", statProps, "targetProps", targetProps)
+	c.info.StatFileProps[cacheKey] = statProps
+	c.info.TargetFileProps[cacheKey] = targetProps
+
+	var (
+		lowestKey     string
+		lowestModTime int
```

Comment:
Time.Unix is guaranteed to return an int64. Could we change lowestModTime to be int64, and initialize it to math.MaxInt64 instead.

Reply:
lovely idea. will incorporate